File: PYTHON/ArrayAndStrings/Easy/BestTimeToBuyAndSellStock.py
# Best Time to Buy and Sell Stock
# Problem: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

def maxProfit(prices):
    """
    Function to calculate the maximum profit from a list of stock prices.
    :param prices: List[int] - List of stock prices where prices[i] is the price of a given stock on day i.
    :return: int - Maximum profit that can be achieved.
    """
    if not prices:
        return 0

    min_price = float('inf')
    max_profit = 0

    for price in prices:
        # Update the minimum price seen so far
        min_price = min(min_price, price)
        # Calculate the profit if sold at the current price
        profit = price - min_price
        # Update the maximum profit
        max_profit = max(max_profit, profit)

    return max_profit

# Example usage
if __name__ == "__main__":
    prices = [7, 1, 5, 3, 6, 4]
    print("Maximum Profit:", maxProfit(prices))  # Output: 5

File: JAVASCRIPT/ArrayAndStrings/Easy/BestTimeToBuyAndSellStock.js
// BEST TIME TO AND SELL STOCK
function maxProfit(prices) {
    let minPrice = Infinity;
    let maxProfit = 0;

    for (let price of prices) {
        if (price < minPrice) {
            minPrice = price;
        } else if (price - minPrice > maxProfit) {
            maxProfit = price - minPrice;
        }
    }

    return maxProfit;
}
// Example usage:
const prices = [7, 1, 5, 3, 6, 4];
console.log(maxProfit(prices)); // Output: 5 (Buy at 1, sell at 6)
// Explanation: The maximum profit is achieved by buying on day 2 (price = 1) and selling on day 5 (price = 6).
// If the prices array is empty or has only one price, no profit can be made.
// Example with no profit
const prices2 = [7, 6, 4, 3, 1];
console.log(maxProfit(prices2)); // Output: 0 (No profit can be made)
// Explanation: In this case, the prices are in descending order, so no profit can be made. 